lib/sycamore/sycamore/transforms/merge_elements.py

```diff
@@ -611,6 +611,195 @@ def postprocess_element(self, elem: Element) -> Element:
+        self.tokenizer = tokenizer
+        self.max_tokens = max_tokens
+        self.merge_across_pages = merge_across_pages
+
+    def preprocess_element(self, element: Element) -> Element:
+        if element.type == "Image" and "summary" in element.properties and "summary" in element.properties["summary"]:
+            element.data["token_count"] = len(self.tokenizer.tokenize(element.properties["summary"]["summary"] or ""))
+        else:
+            element.data["token_count"] = len(self.tokenizer.tokenize(element.text_representation or ""))
+        return element
+
+    def postprocess_element(self, element: Element) -> Element:
+        del element.data["token_count"]
+        return element
+
+    def merge_elements(self, document: Document) -> Document:
+        """Use self.should_merge and self.merge to greedily merge consecutive elements.
+        If the next element should be merged into the last 'accumulation' element, merge it.
+
+        Args:
+            document (Document): A document with elements to be merged.
+
+        Returns:
+            Document: The same document, with its elements merged
+        """
+        if len(document.elements) < 2:
+            return document
+
+        first_header = False
+        for element in document.elements:
+            if element.type in ["Section-header", "Title"]:
+                first_header = True
+                element.data["_header"] = element.text_representation
+            else:
+                if not first_header:
+                    element.data["_header"] = None
+
+        to_merge = [self.preprocess_element(e) for e in document.elements]
+        new_elements = [to_merge[0]]
+        for element in to_merge[1:]:
+            if self.should_merge(new_elements[-1], element):
+                new_elements[-1] = self.merge(new_elements[-1], element)
+            else:
+                new_elements.append(element)
+        document.elements = [self.postprocess_element(e) for e in new_elements]
+        document.elements = [e for e in new_elements if e.type not in ["Section-header", "Title"]]
+        return document
+
+    def should_merge(self, element1: Element, element2: Element) -> bool:
+        # deal with empty elements
+        if (
+            DocumentPropertyTypes.PAGE_NUMBER not in element1.properties
+            or DocumentPropertyTypes.PAGE_NUMBER not in element2.properties
+            or element1.type is None
+            or element2.type is None
+        ):
+            return False
+
+        # DO NOT MERGE across pages
+        if (
+            not self.merge_across_pages
+            and element1.properties[DocumentPropertyTypes.PAGE_NUMBER]
+            != element2.properties[DocumentPropertyTypes.PAGE_NUMBER]
+        ):
+            return False
+
+        if element1.data["token_count"] + 1 + element2.data["token_count"] > self.max_tokens and element2.type not in [
+            "Section-header",
+            "Title",
+        ]:
```

Comment:
This if statement is checking the wrong thing. It should be checking the number of tokens in `element1["header"]`, not the number of tokens in what occasionally is `element1.text_representation`.